src/app/core/services/kafka.service.ts

```diff
@@ -111,146 +112,120 @@ export class KafkaService {
-          .topic(topic)
-          .produce(id, info, payload, (err, res) => {
-            if (err) {
-              console.log(err)
-              return this.cacheAnswers(specs)
-            } else {
-              const cacheKey = specs.kafkaObject.value.time
-              return this.removeAnswersFromCache(cacheKey)
-            }
-          })
-      },
-      error => {
-        console.error(
-          'Could not initiate kafka connection ' + JSON.stringify(error)
-        )
-        return Promise.resolve({ res: 'ERROR' })
-      }
-    )
+  sendToKafka(specs, keySchema, valueSchema, payload) {
+    return this.getKafkaInstance()
+      .then(kafka => new Promise((resolve, reject) => {
+          // NOTE: Kafka connection instance to submit to topic
+          const topic = specs.avsc + '_' + specs.name
+          console.log('Sending to: ' + topic)
+          return kafka
+            .topic(topic)
+            .produce(keySchema, valueSchema, payload, (err, res) => {
+              if (err) {
+                reject(err)
+              } else {
+                resolve(res)
+              }
+            })
+      }))
+      .then(() => this.removeAnswersFromCache(specs.kafkaObject.value.time))
+      .catch(error => {
+        console.error('Could not initiate kafka connection ' + JSON.stringify(error))
+        return this.cacheAnswers(specs)
+          .then(() => ({res: 'ERROR'}))
+      });
   }
 
   cacheAnswers(specs) {
     const kafkaObject = specs.kafkaObject
-    this.storage.get(StorageKeys.CACHE_ANSWERS).then(cache => {
-      console.log('KAFKA-SERVICE: Caching answers.')
-      cache[kafkaObject.value.time] = specs
-      this.storage.set(StorageKeys.CACHE_ANSWERS, cache)
-    })
+    return this.storage.get(StorageKeys.CACHE_ANSWERS)
+      .then(cache => {
+        console.log('KAFKA-SERVICE: Caching answers.')
+        cache[kafkaObject.value.time] = specs
+        return this.storage.set(StorageKeys.CACHE_ANSWERS, cache)
+      })
   }
 
   sendAllAnswersInCache() {
     if (!this.cacheSending) {
       this.cacheSending = !this.cacheSending
       this.sendToKafkaFromCache()
+        .catch(e => console.log('Cache could not be sent.'))
         .then(() => (this.cacheSending = !this.cacheSending))
-        .catch(e => console.log('Cache could not be sent: ' + JSON.stringify(e)))
     }
   }
 
   sendToKafkaFromCache() {
     return this.storage.get(StorageKeys.CACHE_ANSWERS).then(cache => {
-      if (!cache) {
-        return this.storage.set(StorageKeys.CACHE_ANSWERS, {})
-      } else {
-        const promises = Object.entries(cache)
-          .filter(([k, v]) => k)
-          .slice(0, 20)
-          .map(([k, v]) => this.createPayloadAndSend(v))
-        return Promise.all(promises)
-          .then(res => {
-            console.log(res)
-            return res
-          })
-      }
+      const promises = Object.entries(cache)
+        .filter(([k, v]) => k)
+        .slice(0, 20)
```

Comment:
any idea why we slice these 20 items?

Reply:
To limit the number of simultaneous requests to Kafka. We could have another loop that keeps calling this, though.